jetty-util/src/main/java/org/eclipse/jetty/util/Blocking.java

```diff
@@ -0,0 +1,305 @@
+ * <h2>Non shared Runnable</h2>
+ * <pre>
+ *     try(Blocking.Runnable onAction = Blocking.runnable())
+ *     {
+ *         someMethod(onAction);
+ *     }
+ * </pre>
+ *
+ * <h2>Shared Runnable</h2>
+ * <pre>
+ *     Blocking.SharedRunnable shared = new Blocking.Shared();
+ *     // ...
+ *     try(Blocking.Runnable onAction = shared.runnable())
+ *     {
+ *         someMethod(onAction);
+ *     }
+ * </pre>
+ *
+ * <h2>Non shared Callback</h2>
+ * <pre>
+ *     try(Blocking.Callback callback = Blocking.callback())
+ *     {
+ *         someMethod(callback);
+ *     }
+ * </pre>
+ *
+ * <h2>Shared Callback</h2>
+ * <pre>
+ *     Blocking.SharedCallback blocker = new Blocking.Shared();
+ *     // ...
+ *     try(Blocking.Runnable onAction = blocker.callback())
+ *     {
+ *         someMethod(onAction);
+ *     }
+ * </pre>
+ */
+public class Blocking
+{
+    private static final Throwable ACQUIRED = new Throwable()
+    {
+        @Override
+        public synchronized Throwable fillInStackTrace()
+        {
+            return this;
+        }
+    };
+    private static final Throwable SUCCEEDED = new Throwable()
+    {
+        @Override
+        public synchronized Throwable fillInStackTrace()
+        {
+            return this;
+        }
+    };
+
+    public interface Runnable extends java.lang.Runnable, AutoCloseable, Invocable
+    {}
+
+    public static Runnable runnable()
+    {
+        return new Runnable()
+        {
+            final CountDownLatch _complete = new CountDownLatch(1);
+
+            @Override
+            public void run()
+            {
+                _complete.countDown();
+            }
+
+            @Override
+            public InvocationType getInvocationType()
+            {
+                return InvocationType.NON_BLOCKING;
+            }
+
+            @Override
+            public void close() throws IOException
+            {
+                try
+                {
+                    _complete.await();
+                }
+                catch (Throwable t)
+                {
+                    throw IO.rethrow(t);
+                }
+            }
+        };
+    }
+
+    public interface Callback extends org.eclipse.jetty.util.Callback, AutoCloseable, Invocable
+    {}
+
+    public static Callback callback()
+    {
+        return new Callback()
+        {
+            private final CompletableFuture<Throwable> _future = new CompletableFuture<>();
+
+            @Override
+            public InvocationType getInvocationType()
+            {
+                return InvocationType.NON_BLOCKING;
+            }
+
+            @Override
+            public void succeeded()
+            {
+                _future.complete(SUCCEEDED);
+            }
+
+            @Override
+            public void failed(Throwable x)
+            {
+                _future.complete(x == null ? new Throwable() : x);
+            }
+
+            @Override
+            public void close() throws IOException
```

Comment:
This would probably be more useful if it also had a close method with a timeout.

See the `FutureCallback` class which does a similar thing, there is `block()` method for infinite timeout but also a `block(long timeout, TimeUnit unit)` method. Because of this it can be used for the blocking WebSocket sends.

Reply:
I'd prefer to leave that until we actually have a use for it.    Specially for the shared instances, as if you wait for a time and it times out, then what does that mean for the shared instance? Is it still valid ? usable? is it reset?
Tell you what, I'll add it for the single use ones....

Reply:
Ah I can't as they have the same signature returned.   So let's not add them unless there is a real use-case.